Sources/NIO/Selector.swift

```diff
@@ -180,83 +255,39 @@ final class Selector<R: Registration> {
-            events.1.flags = UInt16(EV_ADD)
-        case .none:
-            events.0.flags = UInt16(EV_DELETE)
-            events.1.flags = UInt16(EV_DELETE)
-        }
-
-        var offset: Int = 0
-        var numEvents: Int32 = 2
-
-        if let old = oldInterested {
-            switch old {
-            case .read:
-                if events.1.flags == UInt16(EV_DELETE) {
-                    numEvents -= 1
-                }
-            case .write:
-                if events.0.flags == UInt16(EV_DELETE) {
-                    offset += 1
-                    numEvents -= 1
-                }
-            case .none:
-                // Only discard the delete events
-                if events.0.flags == UInt16(EV_DELETE) {
-                    offset += 1
-                    numEvents -= 1
-                }
-                if events.1.flags == UInt16(EV_DELETE) {
-                    numEvents -= 1
-                }
-            case .all:
-                // No need to adjust anything
-                break
-            }
-        } else {
-            // If its not reregister operation we MUST NOT include EV_DELETE as otherwise kevent will fail with ENOENT.
-            if events.0.flags == UInt16(EV_DELETE) {
-                offset += 1
-                numEvents -= 1
-            }
-            if events.1.flags == UInt16(EV_DELETE) {
-                numEvents -= 1
+        func calculateKQueueChange(event: KQueueEventFilterSet) -> UInt16? {
+            guard differences.contains(event) else {
+                return nil
             }
+            return UInt16(newKQueueFilters.contains(event) ? EV_ADD : EV_DELETE)
         }
 
-        if numEvents > 0 {
+        // Allocated on the stack
+        var events = (kevent(), kevent(), kevent())
+        try selectable.withUnsafeFileDescriptor { fd in
             try withUnsafeMutableBytes(of: &events) { event_ptr in
-                precondition(MemoryLayout<kevent>.size * 2 == event_ptr.count)
-                let ptr = event_ptr.baseAddress?.bindMemory(to: kevent.self, capacity: 2)
-
-                try keventChangeSetOnly(event: ptr!.advanced(by: offset), numEvents: numEvents)
+                precondition(MemoryLayout<kevent>.size * 3 == event_ptr.count)
+                let ptr = event_ptr.baseAddress!.bindMemory(to: kevent.self, capacity: 3)
+
+                var index: Int = 0
+                for (event, filter) in [(KQueueEventFilterSet.read, EVFILT_READ), (.write, EVFILT_WRITE), (.except, EVFILT_EXCEPT)] {
+                    if let flags = calculateKQueueChange(event: event) {
+                        ptr[index].ident = UInt(fd)
+                        ptr[index].filter = Int16(filter)
+                        ptr[index].flags = flags
+                        index += 1
+                    }
+                }
+                if index > 0 {
+                    try keventChangeSetOnly(event: ptr, numEvents: Int32(index))
+                }
```

Comment:
Ok, since we're here lets clean this code up a bit. This method is bizarre and completely fails to explain itself. I'd really love to see this function factored out and moved to a method with a doc comment and a quick explanation of the strategy involved (namely, we know we're only going to need to modify at most 3 events, and these can be passed as a C array, so we can allocate a stack of events and lazily fill them in if needed).

Reply:
+1